File: jenjin-core/src/main/java/com/binarymonks/jj/events/EventBus.java
package com.binarymonks.jj.events;

import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import com.binarymonks.jj.async.Function;

public class EventBus {

    static ObjectMap<String, Object> NULLPARAMS = new ObjectMap<>();

    ObjectMap<String, Listeners> listeners = new ObjectMap<>();

    public void register(String message, Function function) {
        if (!listeners.containsKey(message)) {
            listeners.put(message, new Listeners());
        }
        listeners.get(message).funcListeners.add(function);
    }

    public void register(String message, EventHandler eventHandler) {
        if (!listeners.containsKey(message)) {
            listeners.put(message, new Listeners());
        }
        listeners.get(message).handlers.add(eventHandler);
    }

    public void send(String message, ObjectMap<String, Object> eventParams) {
        if (listeners.containsKey(message)) {
            for (Function func : listeners.get(message).funcListeners) {
                func.call();
            }
            for (EventHandler handler : listeners.get(message).handlers) {
                handler.handle(eventParams);
            }
        }
    }

    public void send(String message) {
        send(message, NULLPARAMS);
    }

    public static class Listeners {
        Array<Function> funcListeners = new Array<>();
        Array<EventHandler> handlers = new Array<>();
    }

}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/events/EventBus.kt
package com.binarymonks.jj.core.events

import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ObjectMap
import com.binarymonks.jj.core.api.EventsAPI

typealias Subscriber = () -> Unit
typealias ParamSubscriber = (params: ObjectMap<String, Any>) -> Unit


class EventBus : EventsAPI {

    internal var listeners = ObjectMap<String, Listeners>()
    private var idCounter = 0

    private fun nextID(): Int {
        idCounter++
        return idCounter
    }

    override fun register(message: String, function: Subscriber): Int {
        if (!listeners.containsKey(message)) {
            listeners.put(message, Listeners())
        }
        val id = nextID()
        listeners.get(message).funcListeners.put(id, function)
        return id
    }

    override fun register(message: String, eventHandler: ParamSubscriber): Int {
        if (!listeners.containsKey(message)) {
            listeners.put(message, Listeners())
        }
        val id = nextID()
        listeners.get(message).handlers.put(id, eventHandler)
        return id
    }

    override fun send(message: String, eventParams: ObjectMap<String, Any>) {
        if (listeners.containsKey(message)) {
            for (func in listeners.get(message).funcListeners.values()) {
                func()
            }
            for (handler in listeners.get(message).handlers.values()) {
                handler(eventParams)
            }
        }
    }

    fun deregister(message: String, registerID: Int) {
        if (listeners.containsKey(message)) {
            listeners[message].funcListeners.remove(registerID)
            listeners[message].handlers.remove(registerID)
        }
    }

    fun clear() {
        listeners.forEach { it.value.funcListeners.clear(); it.value.handlers.clear() }
    }

    class Listeners {
        internal var funcListeners = ObjectMap<Int, Subscriber>()
        internal var handlers = ObjectMap<Int, ParamSubscriber>()
    }

    companion object {

        internal var NULLPARAMS = ObjectMap<String, Any>()
    }

}
